internal/http/services/owncloud/ocdav/delete.go

```diff
@@ -58,49 +60,72 @@ func (s *svc) handlePathDelete(w http.ResponseWriter, r *http.Request, ns string
 }
 
 func (s *svc) handleDelete(ctx context.Context, w http.ResponseWriter, r *http.Request, ref *provider.Reference, log zerolog.Logger) {
+
+	ctx, span := rtrace.Provider.Tracer("reva").Start(ctx, "delete")
+	defer span.End()
+
+	req := &provider.DeleteRequest{Ref: ref}
+
+	// FIXME the lock token is part of the application level protocol, it should be part of the DeleteRequest message not the outgoing context
+	ih, ok := parseIfHeader(r.Header.Get("If"))
+	if ok {
+		if len(ih.lists) == 1 && len(ih.lists[0].conditions) == 1 {
+			req.Opaque = utils.AppendPlainToOpaque(req.Opaque, "lockid", ih.lists[0].conditions[0].Token)
+		}
+	} else if r.Header.Get("If") != "" {
+		w.WriteHeader(http.StatusBadRequest)
+		b, err := errors.Marshal(http.StatusBadRequest, "invalid if header", "")
+		errors.HandleWebdavError(&log, w, b, err)
+		return
+	}
+
 	client, err := s.getClient()
 	if err != nil {
 		log.Error().Err(err).Msg("error getting grpc client")
 		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
-
-	ctx, span := rtrace.Provider.Tracer("reva").Start(ctx, "delete")
-	defer span.End()
-
-	req := &provider.DeleteRequest{Ref: ref}
 	res, err := client.Delete(ctx, req)
 	if err != nil {
 		span.RecordError(err)
 		log.Error().Err(err).Msg("error performing delete grpc request")
 		w.WriteHeader(http.StatusInternalServerError)
 		return
-	} else if res.Status.Code != rpc.Code_CODE_OK {
-		if res.Status.Code == rpc.Code_CODE_NOT_FOUND {
-			w.WriteHeader(http.StatusNotFound)
-			// TODO path might be empty or relative...
-			m := fmt.Sprintf("Resource %v not found", ref.Path)
-			b, err := errors.Marshal(errors.SabredavNotFound, m, "")
-			errors.HandleWebdavError(&log, w, b, err)
-		}
-		if res.Status.Code == rpc.Code_CODE_PERMISSION_DENIED {
-			w.WriteHeader(http.StatusForbidden)
-			// TODO path might be empty or relative...
-			m := fmt.Sprintf("Permission denied to delete %v", ref.Path)
-			b, err := errors.Marshal(errors.SabredavPermissionDenied, m, "")
-			errors.HandleWebdavError(&log, w, b, err)
+	}
+	switch res.Status.Code {
+	case rpc.Code_CODE_OK:
+		w.WriteHeader(http.StatusNoContent)
+	case rpc.Code_CODE_NOT_FOUND:
+		w.WriteHeader(http.StatusNotFound)
+		// TODO path might be empty or relative...
+		m := fmt.Sprintf("Resource %v not found", ref.Path)
+		b, err := errors.Marshal(http.StatusNotFound, m, "")
+		errors.HandleWebdavError(&log, w, b, err)
+	case rpc.Code_CODE_PERMISSION_DENIED:
+		status := http.StatusForbidden
+		if lockID := utils.ReadPlainFromOpaque(res.Opaque, "lockid"); lockID != "" {
+			// http://www.webdav.org/specs/rfc4918.html#HEADER_Lock-Token says that the
+			// Lock-Token value is a Coded-URL. We add angle brackets.
+			w.Header().Set("Lock-Token", "<"+lockID+">")
```

Comment:
Hey @butonic why do you need to wrap the lockid within `<` and `>` ?

Reply:
`Lock-Token` is defined in http://www.webdav.org/specs/rfc4918.html#HEADER_Lock-Token as:
```
Lock-Token = "Lock-Token" ":" Coded-URL
```
`Coded-URL` is defined in http://www.webdav.org/specs/rfc4918.html#HEADER_DAV as:
```
Coded-URL        = "<" absolute-URI ">" 
                     ; No linear whitespace (LWS) allowed in Coded-URL
                     ; absolute-URI defined in RFC 3986, [Section 4.3](https://tools.ietf.org/html/rfc3986#section-4.3)
```
`absolute-URI` is defined in https://tools.ietf.org/html/rfc3986#section-4.3 as:
```
 absolute-URI  = scheme ":" hier-part [ "?" query ]
```

So, AFAIU, the `Lock-Token` must be enclosed in angle brackets.

Reply:
Thanks, fair enough (sorry I should have checked the full specs, not only the top level). It's interesting that in WebDAV parlance a lock payload is expected to be a URL after all, whereas a priori we do not impose any format (and WOPI allows any string). With that, I guess it's fair to say:
- WebDAV endpoints in Reva shall respect this format.
- GRPC endpoints in Reva should use a different header - say `x-reva-lockid` ? - to pass a lock token when necessary, so to be compatible with WOPI-managed locks.
Comments?

Reply:
Interesting thoughts :) some answers/observations:
- The WOPI server is actually not in control of the payload of the lock-id. The WOPI clients (i.e. the editor engines) have their own format as the specs say this is just a string: e.g. MS Office passes a JSON with some internal hashes and UUIDs, Collabora Online uses a single hash. I'd say the CS3 lock should not be constrained by the WebDAV specs - which in this case are just too restrictive, I failed to find an alternate representation that allows a totally random string. Rather, it would mean that a file locked via WOPI will have a lock that cannot be handled via pure WebDAV calls (but can be handled via GRPC).

- It does make sense to have a `LockID string` additional optional argument to all messages that modify a resource, sure. I had the impression we could have preferred a header in order not to modify the current protobuf, but after all we're still at a point where it's OK to modify the API.

By the way, incidentally in WOPI a `DELETE` request can only happen on unlocked files (!) or in other terms locked files are never deleted. Thus a WOPI `DELETE` operation does not carry a lock-id. But OK the semantic does not change, a locked file cannot be deleted if the lock-id does not match (as it can't be modified/renamed etc.).

Reply:
ahh, I was unaware that the clients determine the WOPI Lock ID. From [the docs](https://docs.microsoft.com/en-us/microsoft-365/cloud-storage-partner-program/rest/concepts#lock):
> A lock is also used to store a small bit of temporary data associated with a file. This metadata is called the lock ID and is a string with a maximum length of 1024 ASCII characters (see [note on lock ID lengths](https://docs.microsoft.com/en-us/microsoft-365/cloud-storage-partner-program/rest/concepts#lock-length)). WOPI clients can use this metadata for a variety of purposes, but hosts do not need any knowledge or understanding of the contents of the lock ID. **Hosts must treat it as an opaque string.**

@glpatcern The wopi server is the host side in this description, right? IMO it should prefix the WOPI Lock ID with `opaquelocktoken:` and then make a CS3 `SetLock()` call. When returning the lock it should trim the `opaquelocktoken:` prefix. This allows the wopi locks to appear as `opaquelocktoken:{whatever}` in webdav. Locks set via webdav will get a `urn:uuid:` prefix, which the wopiserver will just return as is to WOPI clients.

Reply:
Indeed that's it, the wopiserver is the "host" in that description. OK so let's use the `opaquelocktoken:` prefix, though WebDAV still expects a UUID as per the specs you linked:
```
OpaqueLockToken-URI = "opaquelocktoken:" UUID [Extension]
```
Yet it's the best approximation we have AFAIU, and it's trivial for the wopiserver to handle that prefix. I'll do that in https://github.com/cs3org/wopiserver/pull/51

Reply:
https://github.com/cs3org/wopiserver/pull/51/commits/2ba5e8f6710eb959a9c87c7eaa99ba3d21fc9700 is the WOPI implementation of the above.

Reply:
The webdav api will store a uuid with a `urn:uuid:` prefix as the lock id. When reading it from the CS3 API it will be dumb and just expose the lock id as is (assuming it it a url).

IMO, all the wopi server needs to do is prefix the `X-WOPI-Lock` a client sends with `opaquelocktoken:` before setting it with `SetLock{}`. When it encounters a `urn:uuid:` prefix it can expose it as the `X-WOPI-Lock` as is. WOPI clients already need to be able to deal with any locks they encounter (msoffice JSON vs. collabora hash).

Reply:
Absolutely, WOPI clients must deal with any "garbage" coming as lock payload and must just ignore them as "not theirs", so the WOPI server would expose WebDAV locks as they are and that's all fine. It's only that in the other way around, WebDAV clients may get `opaquelocktoken:some_random_garbage`, which they should consider as "not theirs" even when `some_random_garbage` is NOT a UUID and a strict WebDAV parser might reject it.

Reply:
anyway, I think we can merge this PR, as it treats the lock id completely opaque.

Reply:
Yes it's fine to merge, my concern was on what you found yourself and I mentioned in my last comment - and that's not related to any (python) implementation: a WebDAV parser may legitimately refuse to parse an `opaquelocktoken:random_stuff` when the random stuff is not UUID, that's all.

But now that you suggest "UUID prefix", what the specs say is to use `UUID [Extension]` and `Extension = path`, where `path` can actually be anything. So I'm now thinking to use for WOPI: `opaquelocktoken:hardcoded_uuid encoded-wopi-lock-id`. And here the encoding is actually necessary as we want something that looks like a path and a JSON can be problematic ;-)

That approach would actually allow the WOPI server to uniquely recognize the WOPI locks from anything else as the UUID would be its signature.